nano/node/election.cpp

```diff
@@ -528,6 +535,83 @@ void nano::election::remove_votes (nano::block_hash const & hash_a)
 	}
 }
 
+void nano::election::remove_block (nano::block_hash const & hash_a)
+{
+	debug_assert (!mutex.try_lock ());
+	debug_assert (status.winner->hash () != hash_a);
+	if (auto existing = last_blocks.find (hash_a); existing != last_blocks.end ())
+	{
+		for (auto i (last_votes.begin ()); i != last_votes.end ();)
+		{
+			if (i->second.hash == hash_a)
+			{
+				i = last_votes.erase (i);
+			}
+			else
+			{
+				++i;
+			}
+		}
+		node.network.publish_filter.clear (existing->second);
+		last_blocks.erase (hash_a);
+	}
+}
+
+bool nano::election::replace_by_weight (nano::unique_lock<std::mutex> & lock_a, nano::block_hash const & hash_a)
+{
+	debug_assert (lock_a.owns_lock ());
+	nano::block_hash replaced_block (0);
+	auto winner_hash (status.winner->hash ());
+	// Sort existing blocks tally
+	std::vector<std::pair<nano::block_hash, nano::uint128_t>> sorted;
+	sorted.reserve (last_tally.size ());
+	std::copy (last_tally.begin (), last_tally.end (), std::back_inserter (sorted));
+	// Sort in ascending order
+	std::sort (sorted.begin (), sorted.end (), [](auto const & left, auto const & right) { return left.second < right.second; });
+	lock_a.unlock ();
+	// Replace if lowest tally is below inactive cache new block weight
+	auto inactive_existing (node.active.find_inactive_votes_cache (hash_a));
+	auto inactive_tally (inactive_existing.status.tally);
+	if (inactive_tally > 0 && sorted.size () < max_blocks)
+	{
+		// If count of tally items is less than 10, remove any block without tally
+		for (auto const & [hash, block] : blocks ())
+		{
+			if (std::find_if (sorted.begin (), sorted.end (), [& hash = hash](auto const & item_a) { return item_a.first == hash; }) == sorted.end () && hash != winner_hash)
+			{
+				replaced_block = hash;
+				break;
+			}
+		}
+	}
+	else if (inactive_tally > 0 && inactive_tally > sorted.front ().second)
+	{
+		if (sorted.front ().first != winner_hash)
+		{
+			replaced_block = sorted.front ().first;
+		}
+		else if (inactive_tally > sorted[1].second)
+		{
+			// Avoid removing winner
```

Comment:
Could the winner change while the mutex is unlocked and another vote is received? Meaning it would assert in `remove_block`

Reply:
Thanks, was fixed in one of the previous commits, but then rolled back again